### ISSUE TITLE
Apt: list of changes

### ISSUE THREAD

**Issue by dmke** [ansible/ansible-modules-core#3801]
##### ISSUE TYPE
- Feature Idea
##### COMPONENT NAME

`apt`
##### ANSIBLE VERSION

```
ansible 2.0.2.0 (stable-2.0 de3f013152) last updated 2016/05/23 14:38:47 (GMT +200)
  lib/ansible/modules/core: (detached HEAD 4b42c6ad59) last updated 2016/05/04 12:18:08 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD 4eb177e545) last updated 2016/04/21 13:02:11 (GMT +200)
  config file = /home/dm/code/admin/ansible/ansible.cfg
  configured module search path = /usr/share/ansible:./library
```
##### CONFIGURATION

Not relevant.
##### OS / ENVIRONMENT
- Managing node: Debian Jessie
- Managed nodes: Debian Jessie, Ubuntu 14.04, Ubuntu 16.04
##### SUMMARY

When running a playbook like this:

``` yml
# playbooks/apt-update.yml
- hosts: all
  tasks:
  - apt: update_cache=yes cache_valid_time=600
  - apt: upgrade=dist
```

no information about the updated packages is displayed. Depending on the updated packages, I need to take further action.

Examples of such actions:
- When our monitoring system gets updated (icinga2), it overwrites the configuration files, and I need to run the Icinga Playbook to fix it.
- When openssl (i.e. libssl) gets updated, I should restart some servers (web, mail and others), so they pick up the new shared library.

My current workaround is to either register a variable with the `apt` task and print the output with `debug: var=apt_result`, or I to inspect `/var/log/apt/history.log`.

It would be nice to have a `show_changes=yes` parameter which either parses the output of `apt`/`aptitude` or reads `/var/log/apt/history.log` and prints the last entry.
##### STEPS TO REPRODUCE

Not a bug.
##### EXPECTED RESULTS

See summary above.
##### ACTUAL RESULTS

See summary above.

**Comment by resmo** [comment]
IMHO this is implemented in 2.1 using --diff, however there is a bug fix #3816 currently waiting to be merged.

**Comment by dmke** [comment]
Thanks for the pointer, but I don't think #3816 is relevant for to this issue.

The [`parse_diff`](https://github.com/ansible/ansible-modules-core/blob/stable-2.1/packaging/os/apt.py#L374) function in stable-2.1's `apt.py` though seems to do what I want, but I fail to see how it gets triggered (there's an [`m._diff`](https://github.com/ansible/ansible-modules-core/blob/stable-2.1/packaging/os/apt.py#L634) conditional, which must be set somewhere...)

**Comment by towolf** [comment]
This was added by my PR https://github.com/ansible/ansible-modules-core/pull/2944

Before Ansible 2.1 there was only the possibility to actually diff 'before' and 'after' texts returned from modules.

With {apt,aptitude,apt-get} that is not so useful. They do provide the dry-run or simulate option, which display what changes would be done (or were done) during a dist-upgrade.

So I added the 'prepared' key to the diff handling, which will just pass-through the text that a module thinks works as a "diff". See https://github.com/ansible/ansible/pull/14105

So, bottom line, you need Ansible 2.1 with the newer apt.py module.

**Comment by towolf** [comment]
@resmo, sorry for the two things that I didn't catch and handle in my original PR, by the way.

**Comment by dmke** [comment]
Indeed, `ansible-playbook playbooks/apt-update.yml --diff` works as advertised, thanks (I didn't notice that Ansible 2.1 was available when I opened this issue).

Is there a way to trigger `--diff` for the `apt` module only, e.g. by adding `diff=True` to the task?

**Comment by towolf** [comment]
There's an open issue to implement what you are asking for: see https://github.com/ansible/ansible/issues/14860

**Comment by dmke** [comment]
Cool, thanks again!
